rfcs/0010-eaglesong/eaglesong.c

```diff
@@ -0,0 +1,199 @@
+0x5c9a812e ,  0x6611fd59 ,  0x85e46044 ,  0x1981d885 ,  0x5a4c903f ,  0x43f30d4b ,  0x7d1d601b ,  0xdd3c3391 ,
+0x030ec65e ,  0xc12878cd ,  0x72e795fe ,  0xd0c76abd ,  0x1ec085db ,  0x7cbb61fa ,  0x93e8dd1e ,  0x8582eb06 ,
+0x73563144 ,  0x049d4e7e ,  0x5fd5aefe ,  0x7b842a00 ,  0x75ced665 ,  0xbb32d458 ,  0x4e83bba7 ,  0x8f15151f ,
+0x7795a125 ,  0xf0842455 ,  0x499af99d ,  0x565cc7fa ,  0xa3b1278d ,  0x3f27ce74 ,  0x96ca058e ,  0x8a497443 ,
+0xa6fb8cae ,  0xc115aa21 ,  0x17504923 ,  0xe4932402 ,  0xaea886c2 ,  0x8eb79af5 ,  0xebd5ea6b ,  0xc7980d3b ,
+0x71369315 ,  0x796e6a66 ,  0x3a7ec708 ,  0xb05175c8 ,  0xe02b74e7 ,  0xeb377ad3 ,  0x6c8c1f54 ,  0xb980c374 ,
+0x59aee281 ,  0x449cb799 ,  0xe01f5605 ,  0xed0e085e ,  0xc9a1a3b4 ,  0xaac481b1 ,  0xc935c39c ,  0xb7d8ce7f };
+
+int num_rounds = 43;
+int capacity = 256;
+int rate = 256;
+
+void PrintState( uint32_t * state ) {
+    int i;
+    for( i = 0 ; i < 16 ; ++i ) {
+        printf("0x%02x%02x%02x%02x ", (state[i] >> (3*8)) & 0xff, (state[i] >> (2*8)) & 0xff, (state[i] >> (1*8)) & 0xff, (state[i] >> (0*8)) & 0xff);
+    }
+
+    printf("\n");
+}
+
+void EaglesongPermutation( uint32_t * state ) {
+    uint32_t new[16];
+    int i, j, k;
+
+    //PrintState(state);
+
+    for( i = 0 ; i < num_rounds ; ++i ) {
+        // bit matrix
+        for( j = 0 ; j < 16 ; ++j ) {
+            new[j] = 0;
+            for( k = 0 ; k < 16 ; ++k ) {
+                new[j] = new[j] ^ (bit_matrix[k*16 + j] * state[k]);
+            }
+        }
+        for( j = 0 ; j < 16 ; ++j ) {
+            state[j] = new[j];
+        }
+
+        // circulant multiplication
+        for( j = 0 ; j < 16 ; ++j ) {
+            state[j] = state[j]  ^  (state[j] << coefficients[3*j+1]) ^ (state[j] >> (32-coefficients[3*j+1]))  ^  (state[j] << coefficients[3*j+2]) ^ (state[j] >> (32-coefficients[3*j+2]));
+        }
+
+        // constants injection
+        for( j = 0 ; j < 16 ; ++j ) {
+            state[j] = state[j] ^ injection_constants[i*16+j];
+        }
+
+        // addition / rotation / addition
+        for( j = 0 ; j < 16 ; j = j + 2 ) {
+            state[j] = state[j] + state[j+1];
+            state[j] = (state[j] << 8) ^ (state[j] >> 24);
+            state[j+1] = (state[j+1] << 24) ^ (state[j+1] >> 8);
+            state[j+1] = state[j] + state[j+1];
+        }
+    }
+}
+
+void EaglesongSponge( unsigned char * output, unsigned int output_length, const unsigned char * input, unsigned int input_length, unsigned char delimiter ) {
+    uint32_t state[16];
+    int i, j, k;
+    uint32_t integer;
+
+    // initialize to zero
+    for( i = 0 ; i < 16 ; ++i ) {
+        state[i] = 0;
+    }
+
+    // absorbing
+    for( i = 0 ; i < ((input_length+1)*8+rate-1) / rate ; ++i ) {
+        for( j = 0 ; j < rate/32 ; ++j ) {
+            integer = 0;
+            for( k = 0 ; k < 4 ; ++k ) {
+                if( i*rate/8 + j*4 + k < input_length ) {
+                    integer = (integer << 8) ^ input[j*4 + k];
```

Comment:
Is the index of input should add i*rate/8?
Now for each loop of i, we fetch same value from input.
Nerver fetch value beyond input[32].

Reply:
test case
# rink @ rink-pc in ~/work/eaglesong [15:31:07] 
$ ./hash
1111111111111111111111111111111111111111111111
76a52594935beb5d00273d9989c0c594da66f45f4e4b22a1907afbcfbaf2f4e2

# rink @ rink-pc in ~/work/eaglesong [15:32:53] 
$ ./hash
1111111111111111111111111111111122222222222222
76a52594935beb5d00273d9989c0c594da66f45f4e4b22a1907afbcfbaf2f4e2

Reply:
Yes, you're right. Great catch! It should be fixed in the most recent commit. For reference:
```
>$ ./hash 
1111111111111111111111111111111111111111111111
d5c1b3ea9c76712bd3da03cc0f330fa5802e25c2f7c7e43dc0fd7db7dbb9f106
>$ ./hash 
1111111111111111111111111111111122222222222222
2e079d6f923344630905faed8efaa8c1b13c501e454e083680903e61da6d18fd
```